### ISSUE TITLE
why shellcheck -x very slow?

### ISSUE THREAD

**Issue by kamontat** [koalaman/shellcheck#1063]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or "online"): 0.4.6
- [ ] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's what shellcheck currently says:
Too slow

#### Here's what I wanted or expected to see:
Faster?

**Comment by nihilus** [comment]
good question, Ive noticed this as well, I guess following the sourced files and then rebuilding the script is what takes time.

**Comment by Wizek** [comment]
One thing that might help you is to run/compile+run shellcheck with profiling enabled. That will give you a .perf file detailing which functions take the longest to run that you can inspect.

If you are unsure how you could do that, this may help: http://book.realworldhaskell.org/read/profiling-and-optimization.html

**Comment by nihilus** [comment]
any progress on this?

**Comment by koalaman** [comment]
Are you saying that checking one script with `-x` is significantly slower than checking all the same scripts without `-x`?

**Comment by kamontat** [comment]
Yes! @koalaman

**Comment by nihilus** [comment]
@koalaman yupp, when you have got about 80 000 LOC this becomes kinda annoying since shellcjeck is the bottleneck in the build environment.

**Comment by kamontat** [comment]
Woww, Your project is every big. But It unnecessarily much like that. My project about 4xxx loc also extremely slow. 
![screen shot 2018-01-19 at 03 13 26](https://user-images.githubusercontent.com/14089557/35119340-3008ed04-fcc7-11e7-9e77-1a970ffe2859.png)

**Comment by koalaman** [comment]
Do either of you have a github repo or similar that I can use as a test case for this particular issue?

I can imagine several reasons why it would happen, but it would be good to have a realistic codebase to test it on.

**Comment by nihilus** [comment]
@koalaman one thing that adds overhead for me is the invokation of proot for chroot-magic for every invokation of shellcheck.

**Comment by koalaman** [comment]
@nihilus Your scripts source by absolute paths that don't correspond to the current system's?

**Comment by nihilus** [comment]
@koalaman that is correct... Most build enviroments for embedded Linux targets a "romfs" root directory in the build enviroment, and from there you create the file system for the flash mtd.

**Comment by rakus** [comment]
I can confirm the bottleneck. In my case it is easy to explain. I have a seven test scripts that test functions that are sourced from another file. Overall my test scripts source the script under test 20 times. 

Checking without `-x`  takes  2200 ms, with `-x` about 10092 ms (all numbers average of 20 executions).

Checking the sourced file with shellcheck takes about 412ms. 

So: `2200 + 20 * 412 = 10440`.  Pretty close.

My case would benefit,  if the sourced file would be cached.

**Comment by balupton** [comment]
For [bevry/dorothy](https://github.com/bevry/dorothy) we've been using shellcheck via [trunk](https://docs.trunk.io/references/cli). However, on our `dev/devilbird` branch which includes a year or updates, and also now has our 200 or so bash commands now sourcing a 5000 line `bash.bash` file, using shellcheck via trunk is now untenably slow. Running `shellcheck commands/{the-command}` works fine, however if I do `shellcheck -x commands/{the-command}` then yeah, it is extremely slow, so I guess Trunk uses the `-x` option. Running `shellcheck commands/setup-util` takes 5 seconds, `shellcheck -x commands/setup-util` takes 3-6 minutes.  I've had to disable the shellcheck linter.

```
$  shellcheck --version
ShellCheck - shell script analysis tool
version: 0.9.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
```

Here's some activity monitor stats for `shellcheck -x commands/setup-util` after just a minute or so of running:

<img width="1692" height="988" alt="Image" src="https://github.com/user-attachments/assets/36151ebd-76a4-44ef-95a5-ba2cda177c76" />

<img width="1692" height="988" alt="Image" src="https://github.com/user-attachments/assets/adb10b8e-baa5-430c-9968-3695eda5fded" />

After two minutes:

<img width="1692" height="988" alt="Image" src="https://github.com/user-attachments/assets/cc8c7d27-5f6b-4a7f-8150-fe7be13debab" />

<img width="1692" height="988" alt="Image" src="https://github.com/user-attachments/assets/ebd90528-e795-4457-af0f-62faaebff48f" />

Okay, version 0.11 is an improvement, taking 2 minutes to complete, with an average memory of 2GB but peaking at 6GB once output starts:

```
$ shellcheck --version
ShellCheck - shell script analysis tool
version: 0.11.0
license: GNU General Public License, version 3
website: https://www.shellcheck.net
```

<img width="1692" height="988" alt="Image" src="https://github.com/user-attachments/assets/58bcdd67-5cb0-42c2-9b78-2edc7eaa38da" />

<img width="1692" height="988" alt="Image" src="https://github.com/user-attachments/assets/7cb2972f-b68e-498f-acf3-0d2d5bbd6af4" />

With `shellcheck commands/setup-util` (without `-x`) still being instant.
